engine/equivalence_source_target.ml

```diff
@@ -35,13 +35,26 @@ let constructor_to_domain repr_env : constructor -> domain = function
      | Tag t -> Domain.tag (Domain.Set.point t)
 
 let constrained_subtrees repr_env children fallback =
+  let head_domain kst_list =
+    let _head_domain : constructor -> Domain.t = function
+      | String _ -> failwith "Not implemented"
+      (* | Variant _v -> assert false *)
+      | Nil -> Domain.empty
+      | Bool _ -> Domain.union (Domain.int (Domain.Set.point 0)) (Domain.int (Domain.Set.point 1))
+      | Int _ -> Domain.int (Domain.Set.full)
+      | Variant _ | Tuple _ | Cons -> Domain.union (Domain.int (Domain.Set.full)) (Domain.tag (Domain.Set.full))
```

Comment:
`Tuple` is `tag 0`: all tuples have tag 0.

For `Variant`, we need to get the list of all constructors of this type, and then use the `repr_env` to get the domain of each constructor. I think that for the first part we need `type_env` also as an extra parameter -- so maybe we need to pass it recursively in the equivalence checker, not just `repr_env`.